src/Interpreters/InterpreterDropQuery.cpp

```diff
@@ -424,18 +425,28 @@ BlockIO InterpreterDropQuery::executeToDatabaseImpl(const ASTDropQuery & query,
                 auto table_context = Context::createCopy(getContext());
                 table_context->setInternalQuery(true);
                 /// Do not hold extra shared pointers to tables
-                std::vector<std::pair<String, bool>> tables_to_drop;
+                std::vector<std::pair<StorageID, bool>> tables_to_drop;
                 // NOTE: This means we wait for all tables to be loaded inside getTablesIterator() call in case of `async_load_databases = true`.
                 for (auto iterator = database->getTablesIterator(table_context); iterator->isValid(); iterator->next())
                 {
                     auto table_ptr = iterator->table();
-                    table_ptr->flushAndPrepareForShutdown();
-                    tables_to_drop.push_back({iterator->name(), table_ptr->isDictionary()});
+                    tables_to_drop.push_back({table_ptr->getStorageID(), table_ptr->isDictionary()});
                 }
 
+                /// Prepare tables for shutdown in parallel.
+                ThreadPoolCallbackRunnerLocal<void> runner(getDatabaseCatalogDropTablesThreadPool().get(), "DropTables");
+                for (const auto & [name, _] : tables_to_drop)
+                {
+                    auto table_ptr = DatabaseCatalog::instance().getTable(name, table_context);
+                    runner([my_table_ptr = std::move(table_ptr)](){
+                        my_table_ptr->flushAndPrepareForShutdown();
+                    });
+                }
+                runner.waitForAllToFinishAndRethrowFirstError();
+
```

Comment:
I found this while was investigating a flaky test. Please take a look here: https://github.com/ClickHouse/ClickHouse/issues/67216#issuecomment-2252986742